include/targets/cuda/kernel.h

```diff
@@ -0,0 +1,80 @@
+#pragma once
+
+#include <target_device.h>
+#include <kernel_helper.h>
+
+namespace quda {
+
+  template <template <typename> class Functor, typename Arg, bool grid_stride = false>
+  __forceinline__ __device__ void Kernel1D_impl(const Arg &arg)
+  {
+    Functor<Arg> f(arg);
+
+    auto i = threadIdx.x + blockIdx.x * blockDim.x;
+
+    while (i < arg.threads.x) {
+      f(i);
+      if (grid_stride) i += gridDim.x * blockDim.x; else break;
+    }
+  }
+
+  template <template <typename> class Functor, typename Arg, bool grid_stride = false>
+    __global__ std::enable_if_t<device::use_kernel_arg<Arg>(), void> Kernel1D(Arg arg) { Kernel1D_impl<Functor, Arg, grid_stride>(arg); }
+
+  template <template <typename> class Functor, typename Arg, bool grid_stride = false>
+    __global__ std::enable_if_t<!device::use_kernel_arg<Arg>(), void> Kernel1D() { Kernel1D_impl<Functor, Arg, grid_stride>(device::get_arg<Arg>()); }
+
+
+  template <template <typename> class Functor, typename Arg, bool grid_stride = false>
+  __forceinline__ __device__ void Kernel2D_impl(const Arg &arg)
+  {
+    Functor<Arg> f(arg);
+
+    auto i = threadIdx.x + blockIdx.x * blockDim.x;
+    auto j = threadIdx.y + blockIdx.y * blockDim.y;
+    if (j >= arg.threads.y) return;
+
+    while (i < arg.threads.x) {
+      f(i, j);
+      if (grid_stride) i += gridDim.x * blockDim.x; else break;
+    }
+  }
+
+  template <template <typename> class Functor, typename Arg, bool grid_stride = false>
+    __global__ std::enable_if_t<device::use_kernel_arg<Arg>(), void> Kernel2D(Arg arg) { Kernel2D_impl<Functor, Arg, grid_stride>(arg); }
+
+  template <template <typename> class Functor, typename Arg, bool grid_stride = false>
+    __global__ std::enable_if_t<!device::use_kernel_arg<Arg>(), void> Kernel2D() { Kernel2D_impl<Functor, Arg, grid_stride>(device::get_arg<Arg>()); }
+
+
+  template <template <typename> class Functor, typename Arg, bool grid_stride = false>
+  __forceinline__ __device__ void Kernel3D_impl(const Arg &arg)
+  {
+    Functor<Arg> f(arg);
+
+    auto i = threadIdx.x + blockIdx.x * blockDim.x;
+    auto j = threadIdx.y + blockIdx.y * blockDim.y;
+    auto k = threadIdx.z + blockIdx.z * blockDim.z;
+    if (j >= arg.threads.y) return;
```

Comment:
Is there already a strategy to prevent deadlocks when using syncs here? Should we  generalize the functor to be something like f(i,j,k,active)? Or is this already compatible with CG?

Reply:
Using `__syncthreads()` is fine in `f(i, j, k)` (on or after Volta), since exited threads will not participate in `__syncthreads()`.

Reply:
This isn't an issue on pre-Volta either.  Exited threads do not contribute to the barrier.